### PR TITLE
Add heuristics to detect assignments whose id has complex type annotations

### LINKED ISSUE #10850: Ugly formatting for complex Map types in Flow and TS
**Prettier 2.3.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBbAhgByUwFkCAeAHSk0wDEBXKMGAS2gBoqbT8yMAnFlADmmAD6YAbhBYATdpmBTcAG3pwiAVSgqIuWXFkAROGjCD8MCP0wBfAHxV7mALyYocAO4kCACgCUANxUIOwgEJZsUGjIoLj8-BCeAArxCDEoqp64AJ4xYQBG-LhgANZwMADK+CVCwsgw-OphcNgFhgayADK4IvS4wnC01ngwrCLIILj0VqEgABYw2CoA6vMs8Gg1YHCV6Rsskhs5k2Bo+SBCaHD8MMnFwnjIAGaq12EAVmgAHgBCxWUKpVcNg4F0hHAXm84J8fpU6io4ABFegQeBQlTvEA1fjXfiTAq4doqOb4QSwFZyGDzZAADgADGEyRBritivhJmTTDdJJCwgBHVHwe4RDJTNAAWg8hkMc34cEFLHl9wGTyQr0xMJA12wLAaTS1aARyKFkPV0LCMCJlNk1OQACZLcUWCo6gBhCDYNUgUwAVjm9GuABUiRkNVjJOoAJJQAywSrmFiWACCscqMByiIx11stiAA)

```sh
--parser babel
```

**Input:**

```jsx
const map: Map<
  Function,
  Map<string | void, { value: UnloadedDescriptor }>
> = new Map();

```

**Output:**

```jsx
const map: Map<Function, Map<string | void, { value: UnloadedDescriptor }>> =
  new Map();

```

**Expected behavior:**
Same as input
```jsx
const map: Map<
  Function,
  Map<string | void, { value: UnloadedDescriptor }>
> = new Map();

```

Since 2.3.0. Feedback from https://github.com/babel/babel/pull/13288/files#r629071597

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #10850

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/print/assignment.js [hunk 1]
```diff
@@ -269,6 +273,48 @@ function isTypeAlias(node) {
...
+              isNonEmptyArray(getTypeParametersFromTypeReference(param)) ||
+              param.type === "TSConditionalType"
+          )
+        ) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
```
Comment (fisker) [inline src/language-js/print/assignment.js:298]: Can we return early?

File: tests/format/flow/assignments/jsfmt.spec.js [hunk 2]
```diff
@@ -0,0 +1 @@
+run_spec(__dirname, ["babel-flow"]);
```
Comment (fisker) [inline tests/format/flow/assignments/jsfmt.spec.js:1]: Run test on `flow` parser too?

File: src/language-js/print/assignment.js [hunk 3]
```diff
@@ -269,6 +273,48 @@ function isTypeAlias(node) {
   return node.type === "TSTypeAliasDeclaration" || node.type === "TypeAlias";
 }
 
+function isComplexTypeAnnotationForAssignments(node) {
+  if (isAssignmentOrVariableDeclarator(node)) {
```
Comment (thorn0) [inline src/language-js/print/assignment.js:277]:
```suggestion
  if (node.type === "VariableDeclarator") {
```

File: src/language-js/print/assignment.js [hunk 4]
```diff
@@ -269,6 +273,48 @@ function isTypeAlias(node) {
   return node.type === "TSTypeAliasDeclaration" || node.type === "TypeAlias";
 }
 
+function isComplexTypeAnnotationForAssignments(node) {
```
Comment (thorn0) [inline src/language-js/print/assignment.js:276]: Let's rename to `hasComplexTypeAnnotation`

File: src/language-js/print/assignment.js [hunk 5]
```diff
@@ -139,7 +139,7 @@ function chooseLayout(path, options, print, leftDoc, rightPropertyName) {
   if (
     isComplexDestructuring(node) ||
     isComplexTypeAliasParams(node) ||
-    isComplexTypeAnnotationForAssignments(node)
+    isComplexTypeAnnotation(node)
```
Comment (thorn0) [inline src/language-js/print/assignment.js:142]: I meant `has...`, not `is...`. The node itself isn't a type annotation.

**Review by fisker — APPROVED** [review]

**Pull request merged**